### ISSUE TITLE
Does a Post Method convertfrom-json automatically?

### ISSUE THREAD

**Issue by bateskevin** [Badgerati/Pode#502]
### Does a Post Method convertfrom-json automatically?

So I'm trying to make the following code work:

```
Start-PodeServer {
    Add-PodeEndpoint -Address localhost -Port 80 -Protocol Http

    Add-PodeRoute -Method Post -Path '/' -ContentType 'application/json' -ScriptBlock {
        
        param($session)
        #$Parameters = $session | ConvertFrom-Json 

        Foreach($Message in $session){
            $Date = Get-Date -f yyyyMMddhhmmss
            $radomnumber = 0..10000 | get-random

            $Name = $Message.Message.Fields.Name

            $Message | ConvertTo-Json | Out-File "C:\Users\USERNAME\Output\$($Date)_$($Name)_$($radomnumber).json"
        }
    }
}
```

Basically just receiving the json that I send with my invoke-restmethod and then adding it to a file with a specific name on the Server where I run pode.

When I try to send a JSON Object for example like this:

JSON:
```
[{
    "messages": [{
            "text": "This is an audit log test",
            "timestamp": 1513702946,
            "fields": [{
                    "name": "SomeField",
                    "content": "Should not be relevant"
                }
            ]
        }
    ]
},
{
    "messages": [{
            "text": "This is an audit log test",
            "timestamp": 1513702946,
            "fields": [{
                    "name": "SomeField",
                    "content": "Should not be relevant"
                }
            ]
        }
    ]
}
]
```

Invoke-RestMethod:
```
Invoke-RestMethod -Uri http://localhost:80/ -Method POST -Body (Get-Content "C:\Users\USERNAME\Message.json")
```

Message.json containing the json above.

Does anybody see an obvious mistake here?

**Comment by Badgerati** [comment]
Hey @bateskevin,

You're right in saying that Pode does auto-convert the Body from JSON (due to the `-ContentType application/json` on the Route).

The `$session` parameter supplied to the Route actually contains a lot of information about the current request, details about it can be [seen here](https://badgerati.github.io/Pode/Tutorials/WebEvent/).

Since you're sending a POST request with a Body, then the JSON will be parsed and available in the `.Data` property of the `$session` (in this case as an array of message objects).

Changing the `foreach` to using `$session.Data` should work:

```powershell
Start-PodeServer {
    Add-PodeEndpoint -Address localhost -Port 80 -Protocol Http

    Add-PodeRoute -Method Post -Path '/' -ContentType 'application/json' -ScriptBlock {
        param($session)

        Foreach($Message in $session.Data){
            $Date = Get-Date -f yyyyMMddhhmmss
            $radomnumber = 0..10000 | get-random

            $Name = $Message.Messages.Fields.Name

            $Message | ConvertTo-Json | Out-File "C:\Users\USERNAME\Output\$($Date)_$($Name)_$($radomnumber).json"
        }
    }
}
```

(I also noticed that `$Message.Message.Fields.Name` possibly should be `$Message.Messages.Fields.Name` instead?)

Hope that helps!

**Comment by bateskevin** [comment]
Awesome!

Right on both accounts. $session.Data did the trick and it was messages indeed :)

Thank you very much for your quick help!!

**Comment by bateskevin** [comment]
hey @Badgerati 

Another question I have would be if I have a variable in the Path to my api like following:

```
Start-PodeServer {
    Add-PodeEndpoint -Address Localhost -Port 80 -Protocol Http

    Add-PodeRoute -Method Post -Path '/api/v1/messages/ingest/:message_id' -ContentType 'application/json' -ScriptBlock {
        
        param($session)
        #$Parameters = $session | ConvertFrom-Json 

        Foreach($Message in $session.data.messages){
            $Date = Get-Date -f yyyyMMddhhmmss
            $radomnumber = 0..10000 | get-random

            $Name = $Message.Messages.Fields.Name

            $Message | ConvertTo-Json -Depth 7 | Out-File "C:\Users\USERNAME\Output\$($Date)_$($Name)_$($radomnumber).json"
        }
    }
}
```

How would the syntax to access the object in the foreach loop there be?

How does the "message_id" come in to play? I use the endpoint to get events forwarded by an event collecting tool. I see, with wireshark, that the api call gets made as expected and the content of the json that is passed to my api is also as I expect it to be, but I'm afraid, that in my scriptblock I'm doing something wrong..

**Comment by Badgerati** [comment]
Hey @bateskevin,

Parameters from the Route's path can be accessed via [`.Parameters`](https://badgerati.github.io/Pode/Tutorials/Routes/Overview/#parameters), ie:
```powershell
$session.parameters.message_id
```

As for the 400, I can't see anything that would obviously trigger it - other than the foreach referencing `$session.data.messages` and then a few lines down there's `$Name = $Message.Messages.Fields.Name` (though I'd expect that to just return null).

To get more info on the error, you can enable error logging to the terminal:
```powershell
Add-PodeEndpoint -Address Localhost -Port 80 -Protocol Http
New-PodeLoggingMethod -Terminal | Enable-PodeErrorLogging
```

If the terminal is something you can't see, then you can also allow a stacktrace to appear on the error page. If you have a `server.psd1` already then setting the following will work, otherwise you can create a `server.psd1` file at the root of your server with the following content:

```powershell
@{
    Web = @{
        ErrorPages = @{
            ShowExceptions = $true
        }
    }
}
```

**Comment by bateskevin** [comment]
Hey @Badgerati,

Cool, I enabled the stacktrace and I found out, that the messages that I receive from the forwarder are zipped. That's why the following is the error:

```
At C:\Program Files\WindowsPowerShell\Modules\Pode\1.6.0\Private\Helpers.ps1:1132 char:44
+                 $Result.Data = ($Content | ConvertFrom-Json)
```

I outputet $Content to a file and I saw that it's just some squiggly signs in there and no actual json.. looking something like this:

```
‹       íUÉnÛ0ýAè¡"Ú)ådÄ]´ŽH‹*0jb³‘Hƒ¤·ù÷R‹+AÑT§|4›ÞŒ8f	Rf3fòãÑTT»*+fb:öcŒ]ä™§¦‚Òá1ˆ’JI9“gFÆ~÷ˆ2Ö À óŒÍ ·S‘²ê™,ëdRûjd)¨ÚÃ$UÚ·+°Üd„ð%SÆ(+¡N¼¼
¾!x.vßu‹†¼Ì(ë–5_øLÏÕv@ïC´æòp–Æ1& V ’çÙ:Ùëí¢åš? ë$Ú­&ø9cyMO´q¢ïZŽ'šà¶gëí
=äwã-êy’¦7”å|-Ót²•
JÏMS¹"s.•
```

Do you have a way of unzipping a request before converting it from json in pode? Or do you know a way in Powershell? I'll try to research it when I get to it and I would be happy to open a PR if the functionality does not exist yet.

**Comment by Badgerati** [comment]
Hey @bateskevin,

That's a lot of squiggly signs! 😂 Unfortuantely Pode doesn't have an inbuilt way to convert from a zip content type, but there is a way you could write your own.

In Pode you can add [custom body parsers](https://badgerati.github.io/Pode/Tutorials/Middleware/Types/BodyParsing/) for content-types. Assuming that the content type could be `application/zip`:

```powershell
Add-PodeBodyParser -ContentType 'application/zip' -ScriptBlock {
    param($body)
    # somehow convert the body into a readable format
    return ($converted_body | ConvertFrom-Json -AsHashtable)
}
```

I'm not too sure how the conversion from the zip stream would happen, as `Expand-Archive` expects a file 🤔 If you manage to work it out, feel free to open a PR for zip content-types!

**Comment by bateskevin** [comment]
Hey @Badgerati,

Just letting you know that I tried to get this to work, but I could not. In Python there is a module to read the gzip stream out where in Powershell I could find different methods to unzip gzip files, but not an incoming string directly.. The idea of writing the string to a file and then unzip it did not work, plus I also think this would not scale at all in my case at least, since I receive 15000 Events per second at times..

In the end I used python for this.. I will close this here. Thanks for all of your support tho! Also I really enjoy pode and I see myself using it for other usecases!
